### PR TITLE
QueryPhaseResultConsumer to call notifyPartialReduce

### PR DESCRIPTION
**Pull request by javanna**
As part of #60275 QueryPhaseResultConsumer ended up calling SearchProgressListener#onPartialReduce directly instead of notifyPartialReduce. That means we don't catch exceptions that may occur while executing the progress listener callback.

This commit fixes the call and adds a test for this scenario.

### CODE REVIEW

**Comment by javanna** [comment]
It would be nice to enforce calling notify* methods at compile time, but the consumer and the listener are on the same package, hence the consumer is free to call the listener protected methods, which need to be protected because the need to be overridden by its subclasses.

File: server/src/test/java/org/elasticsearch/action/search/QueryPhaseResultConsumerTests.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+            querySearchResult.topDocs(new TopDocsAndMaxScore(topDocs, Float.NaN), new DocValueFormat[0]);
+            querySearchResult.setSearchShardTarget(searchShardTarget);
+            querySearchResult.setShardIndex(i);
+            queryPhaseResultConsumer.consumeResult(querySearchResult, partialReduceLatch::countDown);
+        }
+
+        assertEquals(10, searchProgressListener.onQueryResult.get());
+        assertTrue(partialReduceLatch.await(10, TimeUnit.SECONDS));
+        if (onPartialMergeFailure.get() != null) {
+            throw onPartialMergeFailure.get();
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/action/search/QueryPhaseResultConsumerTests.java:120]: should it be `assertNull(onPartialMergeFailure.get())` instead ?

**Review by jimczi — APPROVED** [review]
LGTM, thanks for adding the test.

File: server/src/test/java/org/elasticsearch/action/search/QueryPhaseResultConsumerTests.java (hunk 1 above)
Comment (javanna) [reply server/src/test/java/org/elasticsearch/action/search/QueryPhaseResultConsumerTests.java:120]: yea that's ok too, I just needed to see the stacktrace when this happened :)

**Pull request merged**